### PR TITLE
Simon/client cache2

### PR DESCRIPTION
**Pull request by psi77**
This includes a few things;

1) PWA support using workbox.  This enables a service worker to cache the entire app.
2) A board cache implementation that uses ngx-pwa/local-storage to store the current board set in indexeddb
3) Simple progress reporting whilst loading/parsing/caching board sets

I haven't cached sounds yet.  I'll raise a ticket for that. (See #89)
Also need to URL.revokeObjectURL for the image blob urls but I haven't worked out where to do that yet, so we currently have a very slow memory leak.  Again, I'll put a ticket in for this unless I can think of a way of fixing it as part of this PR. (See #88)

### CODE REVIEW

**Comment by psi77** [comment]
I've added a fix for issue #88 now

File: src/app/progress/progress.component.css [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+::END::LICENCE:: */
+.maxSpace {
+  width: 100%;
+  text-align: center;
+}
+.progressMessage {
+  font-size: small;
+}
+
+.lds-ellipsis {
```
Comment (blaise17) [inline src/app/progress/progress.component.css]: If this isn't angular material it should have camelCased class names please

File: src/app/progress/progress.component.css [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+OVFPlayer is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+You should have received a copy of the GNU General Public License
+along with OVFPlayer.  If not, see <https://www.gnu.org/licenses/>.
+::END::LICENCE:: */
+.maxSpace {
```
Comment (blaise17) [inline src/app/progress/progress.component.css]: Please can we have a blank line betwen each class?

File: src/app/progress/progress.component.css [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+@keyframes lds-ellipsis3 {
+  0% {
+    transform: scale(1);
+  }
+  100% {
+    transform: scale(0);
+  }
+}
+@keyframes lds-ellipsis2 {
```
Comment (blaise17) [inline src/app/progress/progress.component.css]: please can 2 be in between 1 and 3

File: src/app/progress/progress.component.spec.ts [hunk 4]
```diff
@@ -12,17 +12,28 @@ GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with OVFPlayer.  If not, see <https://www.gnu.org/licenses/>.
 ::END::LICENCE:: */
-import { Pipe, PipeTransform } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (blaise17) [reply src/app/progress/progress.component.spec.ts]: how terribly odd

File: src/app/config-page/config-page.component.ts [hunk 5]
```diff
@@ -60,7 +62,18 @@ export class ConfigPageComponent implements OnInit {
     this.configService.speakOnSpeechbarClick = this.speakOnSpeechbarClick;
     this.configService.scanningConfig = this.scanningConfig;
     // TODO: some kind of validation
-    this.router.navigate(['/main']);
+
+    // clear local cache of page to force a refresh
+    this.boardCache.clear().subscribe(
```
[the remark this answers is not in the document]
Comment (psi77) [reply src/app/config-page/config-page.component.ts]: See issue #91

File: src/app/obf-button/obf-button.component.html [hunk 6]
```diff
@@ -20,10 +20,7 @@
      mat-ripple [matRippleDisabled]="!clickHandler"
      (click)="clickHandler && clickHandler(butt)">
   <div class="buttonImageDiv" *ngIf="butt.getImage() as image">
-    <div *ngIf="image.isSVG(); else normalImage" [innerHTML]="image.getSource() | safe">SVG Image</div>
-    <ng-template #normalImage>
-      <img class="buttonImage" [src]="image.getSource()" *ngIf="image.getSource()" />
-    </ng-template>
+    <img #domImage class="buttonImage" [src]="getDataURL()"/>
```
[the remark this answers is not in the document]
Comment (psi77) [reply src/app/obf-button/obf-button.component.html]: It was for some secret code that I'm not going to tell you about

File: src/app/obf-button/obf-button.component.ts [hunk 7]
```diff
@@ -12,23 +12,42 @@ GNU General Public License for more details.
 You should have received a copy of the GNU General Public License
 along with OVFPlayer.  If not, see <https://www.gnu.org/licenses/>.
 ::END::LICENCE:: */
-import { Component, OnInit, Input } from '@angular/core';
+import { Component, OnInit, Input, AfterViewInit, ViewChild, ElementRef, AfterViewChecked, OnDestroy } from '@angular/core';
```
[the remark this answers is not in the document]
Comment (psi77) [reply src/app/obf-button/obf-button.component.ts]: No.  We don't.

**Comment by joereddington** [comment]
Smoke tests done on desktop and kindle fire.  I only saw "Simple progress reporting whilst loading/parsing/caching board sets" for changing the board, rather than initial load.  Is that intentional? (I _think_ I'd prefer it for both...

**Comment by psi77** [comment]
@joereddington the progress reporting should be on both initial load and page change...
It's not up for very long with the caching in place but it definitely fires on mine with break points in place, but it's worth double checking just in case it's only working on my machine...

**Comment by psi77** [comment]
Yes, we squash and merge from feature PRs

**Comment by psi77** [comment]
I've added the last couple of tests that I wanted to add.  Please let me know if you think there is more test coverage that we require.

File: src/app/board-cache.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  });
+
+  it('should be created', inject([BoardCacheService], (service: BoardCacheService) => {
+    expect(service).toBeTruthy();
+  }));
+
+  it(
+    'should call removeItem when clear',
+    inject([BoardCacheService, LocalStorage], (service: BoardCacheService, localStorage: LocalStorage
+  ) => {
```
Comment (blaise17) [inline src/app/board-cache.service.spec.ts]: indentation!

**Pull request merged**